src/aruku/walking/process/kinematic.cpp

```diff
@@ -0,0 +1,682 @@
+  double time, double period, double period_shift, double mag,
+  double mag_shift) const
+{
+  return mag * sin(2_pi / period * time - period_shift) + mag_shift;
+}
+
+bool Kinematic::compute_inverse_kinematic(
+  std::string leg, double x, double y, double z, double a,
+  double b, double c)
+{
+  size_t index = (leg == "right") ? 0 : 6;
+
+  {
+    using keisan::Point3;
+    using keisan::Euler;
+    using keisan::Matrix;
+
+    auto matrix_translation = keisan::translation_matrix(Point3(x, y, z - leg_length));
+    auto matrix_rotation = keisan::rotation_matrix(
+      Euler(
+        keisan::make_radian(a),
+        keisan::make_radian(b), keisan::make_radian(c)));
+
+    auto matrix_transformation = matrix_translation * matrix_rotation;
+
+    auto vector = Point3(
+      x + (matrix_transformation[0][2] * ankle_length),
+      y + (matrix_transformation[1][2] * ankle_length),
+      (z - leg_length) + (matrix_transformation[2][2] * ankle_length));
+
+    // get knee
+    double vector_magnitude = vector.magnitude();
+    double acos_result = acos(
+      ((vector_magnitude * vector_magnitude) - (thigh_length * thigh_length) -
+      (calf_length * calf_length)) / (2 * thigh_length * calf_length));
+
+    if (std::isnan(acos_result)) {
+      return false;
+    }
+    joint_angles[index + 3] = acos_result;
+
+    // get ankle roll
+    auto matrix = matrix_transformation;
+    if (!matrix.inverse()) {
+      return false;
+    }
+
+    double k = sqrt(matrix[1][3] * matrix[1][3] + matrix[2][3] * matrix[2][3]);
+    double l =
+      sqrt(
+      matrix[1][3] * matrix[1][3] + (matrix[2][3] - ankle_length) *
+      (matrix[2][3] - ankle_length));
+    double m = (k * k - l * l - ankle_length * ankle_length) / (2 * l * ankle_length);
+    if (m > 1.0) {
+      m = 1.0;
+    } else if (m < -1.0) {
+      m = -1.0;
+    }
+
+    acos_result = acos(m);
+    if (std::isnan(acos_result)) {
+      return false;
+    }
+    if (matrix[1][3] < 0.0) {
+      joint_angles[index + 5] = -acos_result;
+    } else {
+      joint_angles[index + 5] = acos_result;
+    }
+
+    // get hip yaw
+    matrix_translation = keisan::translation_matrix(Point3(0, 0, ankle_length));
+    matrix_rotation =
+      keisan::rotation_matrix(
+      Euler(
+        keisan::make_radian(joint_angles[index + 5]),
+        keisan::make_radian(0.0), keisan::make_radian(0.0)));
+
+    matrix = matrix_translation * matrix_rotation;
+    if (!matrix.inverse()) {
+      return false;
+    }
+
+    matrix = matrix_transformation * matrix;
+    double atan_result = atan2(-matrix[0][1], matrix[1][1]);
+    if (std::isinf(atan_result)) {
+      return false;
+    }
+    joint_angles[index] = atan_result;
+
+    // get hip roll
+    atan_result =
+      atan2(
+      matrix[2][1],
+      -matrix[0][1] * sin(joint_angles[index] + matrix[1][1] * cos(joint_angles[index])));
+    if (std::isinf(atan_result)) {
+      return false;
+    }
+    joint_angles[index + 1] = atan_result;
```

Comment:
by seing how `joint_angles` is used here, it seems that array doesn't required here, the order doesn't matter, it seems that we can just use our own representation of joint (e.g. `hip_roll = atan result`, or `joint[JointId::HIP_ROLL] = atan_result`).

Reply:
done